### PR TITLE
read write stats and docker image improvements

### PR DESCRIPTION
**Pull request by kinnarr**

### CODE REVIEW

File: src/async_hta_service.hpp [hunk 1]
```diff
@@ -442,6 +450,89 @@ class AsyncHtaService
...
+
+        void decrement_ongoing()
+        {
+            std::scoped_lock lock(stats_lock_);
+            ongoing_requests_count_--;
+            log_stats_();
+        }
+
+    private:
+        void log_stats_()
```
Comment (bmario) [inline src/async_hta_service.hpp]:
```suggestion
        void log_stats()
```

File: src/async_hta_service.hpp [hunk 2]
```diff
@@ -442,6 +450,89 @@ class AsyncHtaService
...
+                    std::chrono::duration_cast<std::chrono::duration<double>>(duration).count();
+            }
+            log_stats_();
+        }
+
+        template <typename T>
+        void add_write_duration(T duration)
+        {
+            {
+                std::scoped_lock lock(stats_lock_);
```
Comment (bmario) [inline src/async_hta_service.hpp]: You don't need a `scoped_lock`, which is required for more than one mutexs. `lock_guard` is die idiomatic tool

File: src/async_hta_service.hpp [hunk 3]
```diff
@@ -442,6 +450,89 @@ class AsyncHtaService
...
+        void increment_ongoing()
+        {
+            std::scoped_lock lock(stats_lock_);
+            ongoing_requests_count_++;
+            log_stats_();
+        }
+
+        void decrement_ongoing()
+        {
+            std::scoped_lock lock(stats_lock_);
```
Comment (bmario) [inline src/async_hta_service.hpp]: Remove the lock and call this function only in the functions `add_write_duration` and `add_read_duration`

File: src/async_hta_service.hpp [hunk 4]
```diff
@@ -442,6 +450,89 @@ class AsyncHtaService
         return it.first->second;
     }
 
+private:
+    class Stats
```
Comment (bmario) [inline src/async_hta_service.hpp]: Move that class into its own header file. There's no need to eat the increase file here.

File: src/async_hta_service.hpp [hunk 5]
```diff
@@ -442,6 +450,89 @@ class AsyncHtaService
...
+                read_duration_ = 0;
+                write_duration_ = 0;
+                read_count_ = 0;
+                write_count_ = 0;
+                last_log_ = metricq::Clock::now();
+            }
+        }
+
+    private:
+        std::recursive_mutex stats_lock_;
```
Comment (bmario) [inline src/async_hta_service.hpp]: I don't think you need a recursive mutex here.

**Review by bmario — CHANGES_REQUESTED** [review]

**Comment by tilsche** [comment]
I'll review once the commends from @bmario are addressed.

File: src/read_write_stats.hpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2019, ZIH,
```
Comment (tilsche) [inline src/read_write_stats.hpp:1]: 2020 - even though I wish it was still 2019

File: src/read_write_stats.hpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+#pragma once
+
+#include "log.hpp"
+
+#include <metricq/types.hpp>
+
+#include <chrono>
+#include <mutex>
+
+class Stats
```
Comment (tilsche) [inline src/read_write_stats.hpp]: I suggest `ReadWriteStats`, just as the file is named.

File: src/read_write_stats.hpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+private:
+    void decrement_ongoing()
+    {
+        ongoing_requests_count_--;
+    }
+
+    void log_stats()
+    {
+        auto duration = metricq::Clock::now() - last_log_;
```
Comment (tilsche) [inline src/read_write_stats.hpp:88]: At this point, we probably want the interval to be configurable... For now at least in the interface of the class.

**Review by tilsche — CHANGES_REQUESTED** [review]
We could split *ongoing* into:

1. *pending* - async routine is posted, but the operation is not started yet in the respective strand (this is currently *ongoing*
2. *ongoing* - incremented right before starting the operation in the strand.

Utilization would use pending + ongoing.

I don't have a strong opinion about the type of duration during collection. My tendency is towards keeping it  `metricq:Duration` and convert at output.

File: src/read_write_stats.hpp (hunk 6 above)
Comment (kinnarr) [reply src/read_write_stats.hpp:1]: But this code is from 2019 😢

Comment (tilsche) [reply src/read_write_stats.hpp:1]: oh. ok fine.

File: Dockerfile [hunk 9]
```diff
@@ -15,11 +20,33 @@ RUN mkdir /home/metricq/metricq-db-hta/build
...
 ENV token=$token
 
 ARG metricq_url=amqp://localhost:5672
 ENV metricq_url=$metricq_url
 
-CMD /home/metricq/metricq-db-hta/build/metricq-db-hta --token $token --server $metricq_url -v
+ARG wait_for_rabbitmq_url=127.0.0.1:5672
+ENV wait_for_rabbitmq_url=$wait_for_rabbitmq_url
+
+CMD  /home/metricq/wait-for-it.sh $wait_for_rabbitmq_url -- /usr/bin/metricq-db-hta --token $token --server $metricq_url -v
```
Comment (bmario) [inline Dockerfile:52]: Why is the waitforit url different from the server url?

File: src/read_write_stats.hpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            read_duration_ = metricq::Duration(0);
+            write_duration_ = metricq::Duration(0);
+            read_count_ = 0;
+            write_count_ = 0;
+            last_log_ = metricq::Clock::now();
+        }
+    }
+
+private:
+    std::mutex stats_lock_;
```
Comment (bmario) [inline src/read_write_stats.hpp]: I‘d rather name this mutex_. A lock holds a mutex.

File: Dockerfile (hunk 9 above)
Comment (kinnarr) [reply Dockerfile:52]: Wait-for-it only accepts host:port as argument and server url also contains protocol, username and password.

**Review by bmario — APPROVED** [review]

**Review by tilsche — APPROVED** [review]

**Pull request merged**